File: dsa-problems/Sorting/solutions/javascript/SRT-012-count-within-threshold-after-self.js
class Solution {
  countWithinThreshold(arr, T) {
    const n = arr.length;
    const counts = new Array(n).fill(0);

    for (let i = 0; i < n; i++) {
      for (let j = i + 1; j < n; j++) {
        if (arr[j] - arr[i] <= T) {
          counts[i]++;
        }
      }
    }

    return counts;
  }
}

const fs = require("fs");

const input = fs.readFileSync(0, "utf8").trim();
if (!input) process.exit(0);
const data = input.split(/\s+/);
let idx = 0;
const n = parseInt(data[idx++], 10);
const t = parseInt(data[idx++], 10);
const arr = [];
for (let i = 0; i < n; i++) {
  arr.push(parseInt(data[idx++], 10));
}
const solution = new Solution();
const result = solution.countWithinThreshold(arr, t);
console.log(result.join(" "));

File: dsa-problems/Sorting/solutions/python/SRT-012-count-within-threshold-after-self.py
def count_within_threshold(arr: list[int], T: int) -> list[int]:
    """For each i, count elements j>i where a[j]-a[i] <= T"""
    n = len(arr)
    counts = [0] * n

    # For each element, count right elements where a[j] - a[i] <= T
    # This means a[j] <= a[i] + T
    for i in range(n):
        for j in range(i + 1, n):
            if arr[j] - arr[i] <= T:
                counts[i] += 1

    return counts

def main():
    n, t = map(int, input().split())
    arr = list(map(int, input().split()))
    result = count_within_threshold(arr, t)
    print(' '.join(map(str, result)))

if __name__ == "__main__":
    main()
